File: swati_app/custom_script/payment_entry/payment_entry.js
frappe.ui.form.on('Payment Entry', {
    custom_get_outstanding_project_wise_invoice: function (frm) {
        if (frm.doc.payment_type == "Receive") {
            let dialog = new frappe.ui.Dialog({
                title: __('Filters'),
                fields: [
                    { fieldname: 'from_date', label: __('From Date'), fieldtype: 'Date' },
                    { fieldname: 'to_date', label: __('To Date'), fieldtype: 'Date' },
                    { fieldname: 'project', label: __('Project'), fieldtype: 'Link', options: 'Project' },
                    { fieldname: 'cost_center', label: __('Cost Center'), fieldtype: 'Link', options: 'Cost Center' },
                    { fieldname: 'allocate_payment_amount', label: __('Allocate Payment Amount'), fieldtype: 'Check', default: 1 }
                ],
                primary_action_label: __('Get Invoices'),
                primary_action: function (values) {
                    frappe.call({
                        method: "swati_app.custom_script.payment_entry.payment_entry.get_outstanding_sales_invoices",
                        args: {
                            company: frm.doc.company,
                            party_type: frm.doc.party_type,
                            party: frm.doc.party,
                            posting_date: frm.doc.posting_date,
                            project: values.project,
                            cost_center: values.cost_center
                        },
                        callback: function (r) {
                            if (r.message && r.message.length) {
                                frm.clear_table("references");

                                r.message.forEach(function (invoice) {
                                    frm.add_child("references", {
                                        reference_doctype: "Sales Invoice",
                                        reference_name: invoice.name,
                                        due_date: invoice.due_date,
                                        total_amount: invoice.grand_total || 0,
                                        outstanding_amount: invoice.outstanding_amount || 0,
                                        account: invoice.debit_to || 0,
                                        allocated_amount: values.allocate_payment_amount ? invoice.outstanding_amount : 0
                                    });
                      
                                });

                                frm.refresh_field("references");
                                frappe.msgprint(__("{0} invoices added to references table.", [r.message.length]));
                            } else {
                                frappe.msgprint(__('No outstanding invoices found.'));
                            }
                        }
                    });
                    dialog.hide();
                }
            });

            dialog.show();
        }
        
                if (frm.doc.payment_type == "Pay") {
                    let dialog = new frappe.ui.Dialog({
                        title: __('Filters'),
                        fields: [
                            { fieldname: 'from_date', label: __('From Date'), fieldtype: 'Date' },
                            { fieldname: 'to_date', label: __('To Date'), fieldtype: 'Date' },
                            { fieldname: 'project', label: __('Project'), fieldtype: 'Link', options: 'Project' },
                            { fieldname: 'cost_center', label: __('Cost Center'), fieldtype: 'Link', options: 'Cost Center' },
                            { fieldname: 'allocate_payment_amount', label: __('Allocate Payment Amount'), fieldtype: 'Check', default: 1 }
                        ],
                        primary_action_label: __('Get Invoices'),
                        primary_action: function (values) {
                            frappe.call({
                                method: "swati_app.custom_script.payment_entry.payment_entry.get_outstanding_purchase_invoices",
                                args: {
                                    company: frm.doc.company,
                                    party_type: frm.doc.party_type,
                                    party: frm.doc.party,
                                    posting_date: frm.doc.posting_date,
                                    project: values.project,
                                    cost_center: values.cost_center
                                },
                                callback: function (r) {
                                    if (r.message && r.message.length) {
                                        frm.clear_table("references");
        
                                        r.message.forEach(function (invoice) {
                                            frm.add_child("references", {
                                                reference_doctype: "Purchase Invoice",
                                                reference_name: invoice.name,
                                                due_date: invoice.due_date,
                                                total_amount: invoice.grand_total || 0,
                                                outstanding_amount: invoice.outstanding_amount || 0,
                                                bill_no :invoice.bill_no || 0,
                                                account: invoice.credit_to || 0,
                                                exchange_rate : invoice.conversion_rate || 0,
                                                allocated_amount: values.allocate_payment_amount ? invoice.outstanding_amount : 0
                                            });
                                        });
        
                                        frm.refresh_field("references");
                                        frappe.msgprint(__("{0} invoices added to references table.", [r.message.length]));
                                    } else {
                                        frappe.msgprint(__('No outstanding invoices found.'));
                                    }
                                }
                            });
                            dialog.hide();
                        }
                    });
        
                    dialog.show();
                } 
            },
                custom_get_outstanding_project_wise_order: function (frm) {
                    if (frm.doc.payment_type == "Receive") {
                        let dialog = new frappe.ui.Dialog({
                            title: __('Filters'),
                            fields: [
                                { fieldname: 'from_date', label: __('From Date'), fieldtype: 'Date' },
                                { fieldname: 'to_date', label: __('To Date'), fieldtype: 'Date' },
                                { fieldname: 'project', label: __('Project'), fieldtype: 'Link', options: 'Project' },
                                { fieldname: 'cost_center', label: __('Cost Center'), fieldtype: 'Link', options: 'Cost Center' },
                                { fieldname: 'allocate_payment_amount', label: __('Allocate Payment Amount'), fieldtype: 'Check', default: 1 }
                            ],
                            primary_action_label: __('Get Invoices'),
                            primary_action: function (values) {
                                frappe.call({
                                    method: "swati_app.custom_script.payment_entry.payment_entry.get_outstanding_sales_orders",
                                    args: {
                                        company: frm.doc.company,
                                        party_type: frm.doc.party_type,
                                        party: frm.doc.party,
                                        posting_date: frm.doc.posting_date,
                                        project: values.project,
                                        cost_center: values.cost_center
                                    },
                                    callback: function (r) {
                                        if (r.message && r.message.length) {
                                            frm.clear_table("references");
            
                                            r.message.forEach(function (invoice) {
                                                frm.add_child("references", {
                                                    reference_doctype: "Sales Order",
                                                    reference_name: invoice.name,
                                                    total_amount: invoice.grand_total || 0,
                                                    outstanding_amount: invoice.grand_total - invoice.advance_paid|| 0,
                                                    allocated_amount: values.allocate_payment_amount ? invoice.outstanding_amount : 0
                                                });
                                  
                                            });
            
                                            frm.refresh_field("references");
                                            frappe.msgprint(__("{0} invoices added to references table.", [r.message.length]));
                                        } else {
                                            frappe.msgprint(__('No outstanding invoices found.'));
                                        }
                                    }
                                });
                                dialog.hide();
                            }
                        });
            
                        dialog.show();
                    } 
                            if (frm.doc.payment_type == "Pay") {
                                let dialog = new frappe.ui.Dialog({
                                    title: __('Filters'),
                                    fields: [
                                        { fieldname: 'from_date', label: __('From Date'), fieldtype: 'Date' },
                                        { fieldname: 'to_date', label: __('To Date'), fieldtype: 'Date' },
                                        { fieldname: 'project', label: __('Project'), fieldtype: 'Link', options: 'Project' },
                                        { fieldname: 'cost_center', label: __('Cost Center'), fieldtype: 'Link', options: 'Cost Center' },
                                        { fieldname: 'allocate_payment_amount', label: __('Allocate Payment Amount'), fieldtype: 'Check', default: 1 }
                                    ],
                                    primary_action_label: __('Get Invoices'),
                                    primary_action: function (values) {
                                        frappe.call({
                                            method: "swati_app.custom_script.payment_entry.payment_entry.get_outstanding_purchase_orders",
                                            args: {
                                                company: frm.doc.company,
                                                party_type: frm.doc.party_type,
                                                party: frm.doc.party,
                                                posting_date: frm.doc.posting_date,
                                                project: values.project,
                                                cost_center: values.cost_center
                                            },
                                            callback: function (r) {
                                                if (r.message && r.message.length) {
                                                    frm.clear_table("references");
                    
                                                    r.message.forEach(function (invoice) {
                                                        frm.add_child("references", {
                                                            reference_doctype: "Purchase Order",
                                                            reference_name: invoice.name,
                                                            total_amount: invoice.grand_total || 0,
                                                            outstanding_amount: invoice.grand_total - invouce.advance_paid || 0,
                                                            exchange_rate : invoice.conversion_rate || 0,
                                                            allocated_amount: values.allocate_payment_amount ? invoice.outstanding_amount : 0
                                                        });
                                                    });
                    
                                                    frm.refresh_field("references");
                                                    frappe.msgprint(__("{0} invoices added to references table.", [r.message.length]));
                                                } else {
                                                    frappe.msgprint(__('No outstanding invoices found.'));
                                                }
                                            }
                                        });
                                        dialog.hide();
                                    }
                                });
                    
                                dialog.show();
                            } 
                        }
                    
                 
        });
     

File: swati_app/custom_script/payment_entry/payment_entry.py
import frappe
from frappe import _
from frappe.utils import getdate

def build_conditions(base_conditions, filters, project=None, cost_center=None):
    if project:
        base_conditions += " AND project = %(project)s"
        filters["project"] = project
    if cost_center:
        base_conditions += " AND cost_center = %(cost_center)s"
        filters["cost_center"] = cost_center
    return base_conditions, filters

@frappe.whitelist()
def get_outstanding_sales_invoices(company, party_type, party, posting_date, project=None, cost_center=None):
    conditions = """
        docstatus = 1
        AND outstanding_amount > 0
        AND company = %(company)s
        AND customer = %(party)s
    """
    filters = {
        "company": company,
        "party": party,
        "posting_date": getdate(posting_date),
    }

    conditions, filters = build_conditions(conditions, filters, project, cost_center)

    return frappe.db.sql(f"""
        SELECT
            name, posting_date, due_date, outstanding_amount, customer, project, debit_to, grand_total
        FROM
            `tabSales Invoice`
        WHERE
            {conditions}
        ORDER BY posting_date ASC
    """, filters, as_dict=True)


@frappe.whitelist()
def get_outstanding_sales_orders(company, party_type, party, posting_date, project=None, cost_center=None):
    conditions = """
        docstatus = 1
        AND advance_paid < grand_total
        AND company = %(company)s
        AND customer = %(party)s
    """
    filters = {
        "company": company,
        "party": party,
        "posting_date": getdate(posting_date),
    }

    conditions, filters = build_conditions(conditions, filters, project, cost_center)

    return frappe.db.sql(f"""
        SELECT
            name, transaction_date, grand_total, customer, project, advance_paid
        FROM
            `tabSales Order`
        WHERE
            {conditions}
        ORDER BY transaction_date ASC
    """, filters, as_dict=True)


@frappe.whitelist()
def get_outstanding_purchase_invoices(company, party_type, party, posting_date, project=None, cost_center=None):
    conditions = """
        docstatus = 1
        AND outstanding_amount > 0
        AND company = %(company)s
        AND supplier = %(party)s
    """
    filters = {
        "company": company,
        "party": party,
        "posting_date": getdate(posting_date),
    }

    conditions, filters = build_conditions(conditions, filters, project, cost_center)

    return frappe.db.sql(f"""
        SELECT
            name, posting_date, due_date, bill_no, conversion_rate,
            outstanding_amount, supplier, project, credit_to,grand_total
        FROM
            `tabPurchase Invoice`
        WHERE
            {conditions}
        ORDER BY posting_date ASC
    """, filters, as_dict=True)


@frappe.whitelist()
def get_outstanding_purchase_orders(company, party_type, party, posting_date, project=None, cost_center=None):
    conditions = """
        docstatus = 1
        AND advance_paid < grand_total
        AND company = %(company)s
        AND supplier = %(party)s
    """
    filters = {
        "company": company,
        "party": party,
        "posting_date": getdate(posting_date),
    }

    conditions, filters = build_conditions(conditions, filters, project, cost_center)

    return frappe.db.sql(f"""
        SELECT
            name, transaction_date, grand_total, supplier, project, advance_paid
        FROM
            `tabPurchase Order`
        WHERE
            {conditions}
        ORDER BY transaction_date ASC
    """, filters, as_dict=True)
